src/Microsoft.CodeAnalysis.Razor/RequiredAttributeParser.cs

```diff
@@ -0,0 +1,304 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Evolution;
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    // Internal for testing
+    internal static class RequiredAttributeParser
```

Comment:
This whole thing should be independent of the builder... The builder should **call** into this class, not the other way around.